CI: operator/pkg/ciliumidentity: TestDeleteUsedCIDIsRecreated: Flakes

## CI failure

```
panic: *resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2.CiliumIdentity].Events() called from /home/runner/work/cilium/cilium/operator/pkg/ciliumidentity/controller_test.go:397 has a broken event handler that did not call Done() before event was garbage collected
                                                                                                                                                                                                                                                                                            
goroutine 17 [running]:                                                                                                                                                                                                                                                                     
github.com/cilium/cilium/pkg/k8s/resource.(*subscriber[...]).processLoop.func1()                                                                                                                                                                                                            
	/home/runner/work/cilium/cilium/pkg/k8s/resource/resource.go:607 +0x6c                                                                                                                                                                                                                      
FAIL	github.com/cilium/cilium/operator/pkg/ciliumidentity	2.046s    
```

The relevant bit of code is 
https://github.com/cilium/cilium/blob/bcd71f6b8a6d57f894dda30e7d97304392103e94/operator/pkg/ciliumidentity/controller_test.go#L397-L403

This breaks out of the loop of consuming events once a Sync event is received, but without cancelling the context of the `Events` call, which means that there's a race - an already allocated, next event could be waiting on writing to the channel and eventually get garbage collected (causing the panic).

The test is generally flaky, reproducible locally with `go test -c ./operator/pkg/ciliumidentity  && stress ./ciliumidentity.test -test.run 'TestDeleteUsedCIDIsRecreated' -test.timeout=10s` - there's multiple, different ways it can fail.

These were introduced recently, in https://github.com/cilium/cilium/pull/34743.

--- a/operator/pkg/ciliumidentity/ciliumendpointslice.go
+++ b/operator/pkg/ciliumidentity/ciliumendpointslice.go
@@ -6,6 +6,7 @@ package ciliumidentity
 import (
 	"context"
 	"strconv"
+	"sync"
 
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	"github.com/cilium/cilium/pkg/logging/logfields"
@@ -14,8 +15,9 @@ import (
 // CID controller also watches for CESs. The only reason is to keep a cache of
 // what CIDs are used in CESs. This is required only when CESs are enabled.
 // CID controller won't delete a CID until it is no longer present in any CESs.
-func (c *Controller) processCiliumEndpointSliceEvents(ctx context.Context) error {
+func (c *Controller) processCiliumEndpointSliceEvents(ctx context.Context, wg *sync.WaitGroup) error {
 	if !c.cesEnabled {
+		wg.Done()
 		return nil
 	}
 
@@ -23,6 +25,8 @@ func (c *Controller) processCiliumEndpointSliceEvents(ctx context.Context) error
 		var idsWithNoCESUsage []int64
 
 		switch event.Kind {
+		case resource.Sync:
+			wg.Done()
 		case resource.Upsert:
 			c.logger.Debug("Got Upsert CES event", logfields.CESName, event.Key.String())
 			idsWithNoCESUsage = c.reconciler.cidUsageInCES.ProcessCESUpsert(event.Object.Name, event.Object.Endpoints)

--- a/operator/pkg/ciliumidentity/controller.go
+++ b/operator/pkg/ciliumidentity/controller.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"fmt"
 	"log/slog"
+	"sync"
 	"time"
 
 	"github.com/cilium/hive/cell"
@@ -159,26 +160,43 @@ func (c *Controller) initializeQueues() {
 		workqueue.RateLimitingQueueConfig{Name: "ciliumidentity_resource"})
 }
 
+// startEventProcessing starts the event processing loop for the Controller.
+// It processes events in the following order:
+// 1. CiliumEndpointSlice, Pod, Namespace events concurrently
+// 2. CiliumIdentity events
+//
+// The function uses a WaitGroup to ensure that each set of events is processed
+// sequentially before moving on to the next.
 func (c *Controller) startEventProcessing() {
+	wg := &sync.WaitGroup{}
+
+	wg.Add(3) // Adding delta for ces, pod, and ns events
+
+	c.jobGroup.Add(
+		job.OneShot("proc-ces-events", func(ctx context.Context, health cell.Health) error {
+			return c.processCiliumEndpointSliceEvents(ctx, wg)
+		}))
+	c.jobGroup.Add(
+		job.OneShot("proc-pod-events", func(ctx context.Context, health cell.Health) error {
+			return c.processPodEvents(ctx, wg)
+		}))
+	c.jobGroup.Add(
+		job.OneShot("proc-ns-events", func(ctx context.Context, health cell.Health) error {
+			return c.processNamespaceEvents(ctx, wg)
+		}))
+
+	wg.Wait()
+
+	wg.Add(1) // Adding cid events
 
 	c.jobGroup.Add(
 		job.OneShot("proc-cid-events", func(ctx context.Context, health cell.Health) error {
-			return c.processCiliumIdentityEvents(ctx)
-		}),
+			return c.processCiliumIdentityEvents(ctx, wg)
+		}))
 
-		job.OneShot("proc-pod-events", func(ctx context.Context, health cell.Health) error {
-			return c.processPodEvents(ctx)
-		}),
-
-		job.OneShot("proc-ces-events", func(ctx context.Context, health cell.Health) error {
-			return c.processCiliumEndpointSliceEvents(ctx)
-		}),
-
-		job.OneShot("proc-ns-events", func(ctx context.Context, health cell.Health) error {
-			return c.processNamespaceEvents(ctx)
-		}),
-	)
+	wg.Wait()
 }
+
 func (c *Controller) initReconciler(ctx context.Context) error {
 	var err error
 	c.reconciler, err = newReconciler(ctx, c.logger, c.clientset, c.namespace, c.pod, c.ciliumIdentity, c.ciliumEndpoint, c.ciliumEndpointSlice, c.cesEnabled, c)
@@ -186,11 +204,6 @@ func (c *Controller) initReconciler(ctx context.Context) error {
 		return fmt.Errorf("cid reconciler failed to init: %w", err)
 	}
 	c.logger.Info("Starting CID controller reconciler")
-
-	// The desired state needs to be calculated before the events are processed.
-	if err := c.reconciler.calcDesiredStateOnStartup(); err != nil {
-		return fmt.Errorf("cid controller failed to calculate the desired state: %w", err)
-	}
 	return nil
 }
 

--- a/operator/pkg/ciliumidentity/controller_test.go
+++ b/operator/pkg/ciliumidentity/controller_test.go
@@ -15,6 +15,9 @@ import (
 	"github.com/cilium/hive/job"
 	"github.com/google/go-cmp/cmp"
 	prometheustestutil "github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/cilium/cilium/operator/k8s"
@@ -438,8 +441,6 @@ func TestUpdateUsedCIDIsReverted(t *testing.T) {
 }
 
 func TestDeleteUsedCIDIsRecreated(t *testing.T) {
-	t.Skip("test disabled, see https://github.com/cilium/cilium/issues/35135")
-
 	ns1 := testCreateNSObj("ns1", nil)
 	pod1 := testCreatePodObj("pod1", "ns1", testLbsC, nil)
 	cid1 := testCreateCIDObjNs("1000", pod1, ns1)
@@ -472,59 +473,35 @@ func TestDeleteUsedCIDIsRecreated(t *testing.T) {
 		t.Fatalf("unexpected error while getting CID store: %s", err)
 	}
 
-	var (
-		lastErr  error
-		toDelete *capi_v2.CiliumIdentity
-	)
-	if err := testutils.WaitUntil(func() bool {
+	var cid *capi_v2.CiliumIdentity
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
 		cids := store.List()
+		assert.Len(ct, cids, 1)
 		if len(cids) != 1 {
-			lastErr = fmt.Errorf("expected 1 identity, got %d", len(cids))
-			return false
+			return
 		}
-		toDelete = cids[0]
-		return true
-	}, WaitUntilTimeout); err != nil {
-		t.Fatalf("timeout waiting for identities in store: %s", lastErr)
-	}
-
-	// Start listening to identities events but discard all events being replayed.
-	events := (*cidResource).Events(ctx)
-	for ev := range events {
-		ev.Done(nil)
-		if ev.Kind == resource.Sync {
-			break
+		cid = cids[0]
+		if !cmp.Equal(cid.SecurityLabels, cid1.SecurityLabels) {
+			t.Fatalf("expected labels %v, got %v", cid.SecurityLabels, cid1.SecurityLabels)
 		}
-	}
+	}, WaitUntilTimeout, 100*time.Millisecond)
 
-	// Update identity.
-	if err := fakeClient.CiliumV2().CiliumIdentities().Delete(ctx, toDelete.Name, metav1.DeleteOptions{}); err != nil {
-		t.Fatalf("delete CID: %v", err)
-	}
+	err = fakeClient.CiliumV2().CiliumIdentities().Delete(ctx, cid.Name, metav1.DeleteOptions{})
+	assert.NoError(t, err, "CiliumIdentity deletion should not return an error")
 
-	// Wait for delete event to propagate.
-	ev := <-events
-	if ev.Kind != resource.Delete {
-		t.Fatalf("expected delete event, got %v", ev.Kind)
-	}
-	if ev.Key.Name != toDelete.Name {
-		t.Fatalf("expected delete event for cid %s, got %v", toDelete.Name, ev.Key.Name)
-	}
-	if !cmp.Equal(ev.Object.SecurityLabels, toDelete.SecurityLabels) {
-		t.Fatalf("expected labels %v, got %v", toDelete.SecurityLabels, ev.Object.SecurityLabels)
-	}
-	ev.Done(nil)
+	_, err = fakeClient.CiliumV2().CiliumIdentities().Get(ctx, cid.Name, metav1.GetOptions{})
+	assert.True(t, errors.IsNotFound(err), "Expected NotFound error after deletion")
 
-	// Wait for reconciler to recreate the CID.
-	ev = <-events
-	if ev.Kind != resource.Upsert {
-		t.Fatalf("expected upsert event, got %v", ev.Kind)
-	}
-	if ev.Key.Name != toDelete.Name {
-		t.Fatalf("expected upsert event for cid %s, got %v", toDelete.Name, ev.Key.Name)
-	}
-	if !cmp.Equal(ev.Object.SecurityLabels, cid1.SecurityLabels) {
-		t.Fatalf("expected labels %v, got %v", cid1.SecurityLabels, ev.Object.SecurityLabels)
-	}
-	ev.Done(nil)
+	// Ensure the identity will be re-created
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		cids := store.List()
+		assert.Len(ct, cids, 1)
+		if len(cids) != 1 {
+			return
+		}
+		cid = cids[0]
+		if !cmp.Equal(cid.SecurityLabels, cid1.SecurityLabels) {
+			t.Fatalf("expected labels %v, got %v", cid.SecurityLabels, cid1.SecurityLabels)
+		}
+	}, WaitUntilTimeout, 100*time.Millisecond)
 }

--- a/operator/pkg/ciliumidentity/identity.go
+++ b/operator/pkg/ciliumidentity/identity.go
@@ -5,6 +5,7 @@ package ciliumidentity
 
 import (
 	"context"
+	"sync"
 
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	"github.com/cilium/cilium/pkg/logging/logfields"
@@ -25,13 +26,18 @@ func (c CIDItem) Key() resource.Key {
 func (c CIDItem) Reconcile(reconciler *reconciler) error {
 	return reconciler.reconcileCID(c.key)
 }
+
 func (c CIDItem) Meter(enqueuedLatency float64, processingLatency float64, isErr bool, metrics *Metrics) {
 	metrics.meterLatency(LabelValueCID, enqueuedLatency, processingLatency)
 	metrics.markEvent(LabelValueCID, isErr)
 }
 
-func (c *Controller) processCiliumIdentityEvents(ctx context.Context) error {
+func (c *Controller) processCiliumIdentityEvents(ctx context.Context, wg *sync.WaitGroup) error {
 	for event := range c.ciliumIdentity.Events(ctx) {
+		if event.Kind == resource.Sync {
+			wg.Done()
+		}
+
 		if event.Kind == resource.Upsert || event.Kind == resource.Delete {
 			c.logger.Debug("Got CID event", logfields.Type, event.Kind, logfields.CIDName, event.Key.String())
 			c.enqueueReconciliation(CIDItem{cidResourceKey(event.Object.Name)}, 0)

--- a/operator/pkg/ciliumidentity/namespace.go
+++ b/operator/pkg/ciliumidentity/namespace.go
@@ -5,6 +5,7 @@ package ciliumidentity
 
 import (
 	"context"
+	"sync"
 
 	ciliumio "github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
 	"github.com/cilium/cilium/pkg/k8s/resource"
@@ -19,9 +20,11 @@ func nsResourceKey(namespace string) resource.Key {
 	return resource.Key{Name: namespace}
 }
 
-func (c *Controller) processNamespaceEvents(ctx context.Context) error {
+func (c *Controller) processNamespaceEvents(ctx context.Context, wg *sync.WaitGroup) error {
 	for event := range c.namespace.Events(ctx) {
 		switch event.Kind {
+		case resource.Sync:
+			wg.Done()
 		case resource.Upsert:
 			c.logger.Debug("Got Upsert Namespace event", logfields.K8sNamespace, event.Key.String())
 

--- a/operator/pkg/ciliumidentity/pod.go
+++ b/operator/pkg/ciliumidentity/pod.go
@@ -5,6 +5,7 @@ package ciliumidentity
 
 import (
 	"context"
+	"sync"
 
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	"github.com/cilium/cilium/pkg/logging/logfields"
@@ -30,8 +31,13 @@ func (p PodItem) Meter(enqueuedLatency float64, processingLatency float64, isErr
 	metrics.meterLatency(LabelValuePod, enqueuedLatency, processingLatency)
 	metrics.markEvent(LabelValuePod, isErr)
 }
-func (c *Controller) processPodEvents(ctx context.Context) error {
+
+func (c *Controller) processPodEvents(ctx context.Context, wg *sync.WaitGroup) error {
 	for event := range c.pod.Events(ctx) {
+		if event.Kind == resource.Sync {
+			wg.Done()
+		}
+
 		if event.Kind == resource.Upsert || event.Kind == resource.Delete {
 			c.logger.Debug("Got Pod event", logfields.Type, event.Kind, logfields.K8sPodName, event.Key.String())
 			c.enqueueReconciliation(PodItem{podResourceKey(event.Object.Name, event.Object.Namespace)}, 0)

--- a/operator/pkg/ciliumidentity/reconciler.go
+++ b/operator/pkg/ciliumidentity/reconciler.go
@@ -111,36 +111,6 @@ func newReconciler(
 	return r, nil
 }
 
-// syncCESsOnStartup updates the cache of CID usage in CES for all the
-// existing CESs.
-func (r *reconciler) calcDesiredStateOnStartup() error {
-	r.syncCESsOnStartup()
-	return r.syncPodsOnStartup()
-}
-
-func (r *reconciler) syncCESsOnStartup() {
-	if !r.cesEnabled {
-		return
-	}
-
-	for _, ces := range r.cesStore.List() {
-		r.cidUsageInCES.ProcessCESUpsert(ces.Name, ces.Endpoints)
-	}
-}
-
-// syncPodsOnStartup ensures that all pods have a CID for their labels.
-func (r *reconciler) syncPodsOnStartup() error {
-	var lastError error
-
-	for _, pod := range r.podStore.List() {
-		if err := r.reconcilePod(podResourceKey(pod.Name, pod.Namespace)); err != nil {
-			lastError = err
-		}
-	}
-
-	return lastError
-}
-
 // reconcileCID ensures that the desired state for the CID is reached, by
 // comparing the CID in desired state cache and watcher's store and doing one of
 // the following:

--- a/operator/pkg/ciliumidentity/reconciler_test.go
+++ b/operator/pkg/ciliumidentity/reconciler_test.go
@@ -18,7 +18,6 @@ import (
 	k8sTesting "k8s.io/client-go/testing"
 
 	"github.com/cilium/cilium/operator/k8s"
-	cestest "github.com/cilium/cilium/operator/pkg/ciliumendpointslice/testutils"
 	"github.com/cilium/cilium/pkg/hive"
 	"github.com/cilium/cilium/pkg/identity/key"
 	capi_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
@@ -171,122 +170,6 @@ func testVerifyCIDUsageInPods(t *testing.T, usage *CIDUsageInPods, expectedCIDs,
 
 	if expectedPods != totalUsedCIDs {
 		t.Errorf("Total CID usage does not match expected pod count. Expected: %d, Got: %d", expectedPods, totalUsedCIDs)
-	}
-}
-
-func TestSyncCESsOnStartup(t *testing.T) {
-	testCases := []struct {
-		name          string
-		cesEnabled    bool
-		endpointSlice *capi_v2a1.CiliumEndpointSlice
-		expectedCIDs  map[int64]int
-	}{
-		{
-			name:       "ces_disabled",
-			cesEnabled: false,
-			endpointSlice: cestest.CreateStoreEndpointSlice("ces", "ns",
-				[]capi_v2a1.CoreCiliumEndpoint{
-					cestest.CreateManagerEndpoint("cep1", 1000),
-					cestest.CreateManagerEndpoint("cep2", 1000),
-					cestest.CreateManagerEndpoint("cep3", 2000),
-					cestest.CreateManagerEndpoint("cep4", 2000)}),
-			expectedCIDs: map[int64]int{},
-		},
-		{
-			name:       "ces_enabled",
-			cesEnabled: true,
-			endpointSlice: cestest.CreateStoreEndpointSlice("ces", "ns",
-				[]capi_v2a1.CoreCiliumEndpoint{
-					cestest.CreateManagerEndpoint("cep1", 1000),
-					cestest.CreateManagerEndpoint("cep2", 1000),
-					cestest.CreateManagerEndpoint("cep3", 2000),
-				}),
-			expectedCIDs: map[int64]int{
-				1000: 2,
-				2000: 1,
-			},
-		},
-	}
-
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
-			ctx := context.Background()
-			reconciler, _, _, cleanupFunc := testNewReconciler(t, ctx, tc.cesEnabled)
-			defer cleanupFunc()
-
-			if err := reconciler.cesStore.CacheStore().Add(tc.endpointSlice); err != nil {
-				t.Errorf("Error adding CiliumEndpointSlice to cache store: %v", err)
-			}
-
-			reconciler.syncCESsOnStartup()
-
-			if !reflect.DeepEqual(reconciler.cidUsageInCES.cidUsageCount, tc.expectedCIDs) {
-				t.Errorf("Expected CID usage count map to be %v, but got %v", tc.expectedCIDs, reconciler.cidUsageInCES.cidUsageCount)
-			}
-		})
-	}
-}
-
-func TestSyncPodsOnStartup(t *testing.T) {
-	ns1 := testCreateNSObj("ns1", nil)
-	ns2 := testCreateNSObj("ns2", nil)
-
-	pod1 := testCreatePodObj("pod1", ns1.Name, testLbsA, nil)
-	pod2 := testCreatePodObj("pod2", ns1.Name, testLbsB, nil)
-	pod3 := testCreatePodObj("pod3", ns2.Name, testLbsC, nil)
-
-	testCases := []struct {
-		name          string
-		pods          []*slim_corev1.Pod
-		existingCIDs  []*capi_v2.CiliumIdentity
-		expectedQSize int // Expected size of the queue after sync
-	}{
-		{
-			name: "new_pods",
-			pods: []*slim_corev1.Pod{
-				pod1,
-				pod2,
-				pod3,
-			},
-			expectedQSize: 3,
-		},
-		{
-			name: "cid_with_new_pod",
-			pods: []*slim_corev1.Pod{
-				pod1,
-				pod2,
-				pod3, // new pod
-			},
-			existingCIDs: []*capi_v2.CiliumIdentity{
-				testCreateCIDObjNs("1000", pod1, ns1),
-				testCreateCIDObjNs("2000", pod2, ns1),
-			},
-			expectedQSize: 1,
-		},
-	}
-
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
-			ctx := context.Background()
-			reconciler, queueOps, _, cleanupFunc := testNewReconciler(t, ctx, false)
-			defer cleanupFunc()
-
-			for _, pod := range tc.pods {
-				_ = reconciler.nsStore.CacheStore().Add(testCreateNSObj(pod.Namespace, nil))
-				_ = reconciler.podStore.CacheStore().Add(pod)
-			}
-			for _, cid := range tc.existingCIDs {
-				_ = reconciler.cidStore.CacheStore().Add(cid)
-			}
-
-			if err := reconciler.syncPodsOnStartup(); err != nil {
-				t.Errorf("Error syncing pods on startup: %v", err)
-			}
-
-			if len(queueOps.fakeWorkQueue) != tc.expectedQSize {
-				t.Errorf("Expected queue size to be %d, but got %d", tc.expectedQSize, len(queueOps.fakeWorkQueue))
-			}
-		})
 	}
 }
 

